doc: suit: Push scenario and DFU caches documentation (#17519)
nrfconnect/sdk-nrf: doc/nrf/app_dev/device_guides/working_with_nrf/nrf54h/ug_nrf54h20_suit_external_memory.rst

* IPC service exposed towards the Secure Domain

When the companion image is booted and a directive that accesses the data on the external memory is issued, such as the ``suit-directive-fetch`` or ``suit-directive-copy`` directives, the Secure Domain firmware uses the IPC service provided by the companion image to access the contents of the external memory.

> Beyond the booting of the companion image, the update process does not differ from regular push model-based or fetch model-based updates.

Reviewer: ```suggestion
After booting the companion image, the update process works the same as for regular push or fetch model updates.
```

Author: I'd rather change it to "apart from", not after, it is not about the sequence of operations.

Author: Changed to apart from - see if it's OK

Difference between push and fetch models
========================================

Push model
----------

In the push model, the cache partition contents are created on the building machine and pushed to the device without modifications.
The images are extracted to the cache partition files using the :kconfig:option:`CONFIG_SUIT_DFU_CACHE_EXTRACT_IMAGE`  Kconfig option.

> No adding of additional sequence is needed in the SUIT manifest.

Reviewer: ```suggestion
No additional sequences are required in the SUIT manifest.
```

For more details, see :ref:`ug_nrf54h20_suit_push`.

Fetch model
-----------

> In the fetch model, the SUIT processor is run on the application core.

Reviewer: ```suggestion
In the fetch model, the SUIT processor runs on the application core.
```

> In the SUIT manifest, you can define a component representing the cache partition in the external memory.

Reviewer: ```suggestion
In the SUIT manifest, you can define a component that represents the cache partition in the external memory.
```

> Within the ``suit-payload-fetch`` sequence, you can then store fetched payload(s) into a ``CACHE_POOL`` component.

Reviewer: ```suggestion
Within the ``suit-payload-fetch`` sequence, you can then store fetched payloads into a ``CACHE_POOL`` component.
```

> This way, the device will pull the images from the outside world and manage storing them into the cache partitions.

Reviewer: ```suggestion
The device then pulls the images from external sources and manages their storage in the cache partitions.
```

For more details, see :ref:`ug_nrf54h20_suit_fetch`.

Enabling external flash support in SUIT DFU
*******************************************

> The :ref:`nrf54h_suit_sample` sample contains several premade configurations that enable the external memory in SUIT DFU.

Reviewer: ```suggestion
The :ref:`nrf54h_suit_sample` sample contains several example configurations that enable the external memory for SUIT DFU.
```

> The configurations using the push model are:

Reviewer: ```suggestion
The configurations using the push model are the following:
```

* ``sample.suit.smp_transfer.cache_push.extflash``
* ``sample.suit.smp_transfer.cache_push.extflash.bt``

> The configurations using the fetch model are:

Reviewer: ```suggestion
The configurations using the fetch model are following:
```

* ``sample.suit.smp_transfer.full_processing.extflash``
* ``sample.suit.smp_transfer.full_processing.extflash.bt``

> You can look at the definitions of these configurations in the :file:`samples/suit/smp_transfer/sample.yaml` file to see which options have to be enabled.

Reviewer: ```suggestion
You can find these configurations defined in the :file:`samples/suit/smp_transfer/sample.yaml` file.
This file specifies which options need to be enabled.
```

> Alternatively, below is the description of the steps needed to enable the external memory in SUIT DFU.

Reviewer: ```suggestion
Alternatively, you can follow the following steps to manually enable external memory in SUIT DFU.
```

      In this example, the component index is ``3``.
      In the following steps, the companion image component is selected with ``suit-directive-set-component-index: 3``.

>    #. Modify the ``suit-install`` sequence in the application manifest file :file:`app_envelope.yaml.jinja2` to boot the companion image before accessing the candidate images, which are stored in the external memory:

Reviewer: ```suggestion
   #. Modify the ``suit-install`` sequence in the application manifest file (:file:`app_envelope.yaml.jinja2`) to boot the companion image before accessing the candidate images stored in the external memory:
```

      The companion image can be optionally upgraded and have its integrity checked.

Steps specific for the push model
=================================

> 1. Enable the :kconfig:option:`CONFIG_SUIT_DFU_CANDIDATE_PROCESSING_PUSH_TO_CACHE` to modifications of the cache partitions by the application core.

Author: I forgot the word "enable" ("to enable modifications of cache partitions"). The suggestion changes the intended meaning.
I've added the fix

> 1. Enable the :kconfig:option:`CONFIG_SUIT_DFU_CANDIDATE_PROCESSING_PUSH_TO_CACHE` to enable modifications of the cache partitions by the application core.

Reviewer: ```suggestion
1. Enable the :kconfig:option:`CONFIG_SUIT_DFU_CANDIDATE_PROCESSING_PUSH_TO_CACHE` option to allow the application core to modify cache partitions.
```

> #.  Enable the :kconfig:option:`CONFIG_SUIT_DFU_CACHE_EXTRACT_IMAGE` Kconfig option for every image that needs to be updated from external memory.

Reviewer: ```suggestion
#.  Enable the :kconfig:option:`CONFIG_SUIT_DFU_CACHE_EXTRACT_IMAGE` Kconfig option for each image that must be updated from external memory.
```

Author: must is a too strong wording for me, is "for each image that you want to" ok?

#. Modify the manifest files for all domains by completing the following steps:

>    a. Ensure that the URI used by the ``suit-payload-fetch`` sequence to fetch a given image matches the :kconfig:option:`CONFIG_SUIT_DFU_CACHE_EXTRACT_IMAGE_URI` Kconfig option.

Reviewer: ```suggestion
   a. Ensure that the URI used by the ``suit-payload-fetch`` sequence (to fetch a given image) matches the :kconfig:option:`CONFIG_SUIT_DFU_CACHE_EXTRACT_IMAGE_URI` Kconfig option.
```

Author: This is not the meaning I inteded - I wanted to put the focus on the URI - it is about the URI representing a given image.

Maybe something like this would be better:

" Ensure that the URI used to fetch a given image (passed as a parameter to the ``suit-payload-fetch`` sequence) matches..."

What do you think?

>    #. Ensure that the envelope integrates the given image inside the envelope integrated payloads section.

Reviewer: ```suggestion
   #. Ensure that the envelope integrates the specified image within the envelope integrated payloads section.
```

>       This is ensured by default if using the provided default SUIT envelope templates.

Reviewer: ```suggestion
      This is ensured by default if you use the provided SUIT envelope templates.
```

Steps specific for the fetch model
==================================

> 1. Enable the :kconfig:option:`CONFIG_SUIT_DFU_CANDIDATE_PROCESSING_FULL` Kconfig option to enable processing of the SUIT manifests by the application core.

Reviewer: ```suggestion
1. Enable the :kconfig:option:`CONFIG_SUIT_DFU_CANDIDATE_PROCESSING_FULL` Kconfig option to allow the application core to process SUIT manifests.
```

> #. Enable the :kconfig:option:`CONFIG_SUIT_STREAM_SOURCE_FLASH` Kconfig option, which enables SUIT processor on the application core to read and parse DFU cache partitions.

Reviewer: ```suggestion
#. Enable the :kconfig:option:`CONFIG_SUIT_STREAM_SOURCE_FLASH` Kconfig option, which allows the SUIT processor on the application core to read and parse DFU cache partitions.
```

#. Modify the application manifest file :file:`app_envelope.yaml.jinja2` by completing the following steps:

   a. Modify the ``CACHE_POOL`` identifier in the SUIT manifest:

      .. code-block:: yaml

         suit-components:
             ...
         - - CACHE_POOL
           - 1

      The ``CACHE_POOL`` identifier must match the identifier of the cache partition defined in the DTS file.

   #. Add the ``suit-payload-fetch`` sequence:

      .. code-block:: yaml

         suit-payload-fetch:
         - suit-directive-set-component-index: 2
         - suit-directive-override-parameters:
             suit-parameter-uri: 'file://{{ app['binary'] }}'
         - suit-directive-fetch:
           - suit-send-record-failure

>       This is assuming that `CACHE_POOL` is the third on the components list and is assigned th index 2.

Author: The sentence is supposed to be an explanation where the number 2 in  `suit-directive-set-component-index: 2` comes from in the code snippet above, and not explaining that the third component on the list of components is assigned the index 2 (this should be clear to clients which have already worked on manifests, and this is a page on a more advanced topic), so the rephrasing does not convey what I've intended.

I've modified the sentence to:

"The assumption in the snippet above is that that ``CACHE_POOL`` is the third component on the manifest's components list (so its component index is 2)."

>       The assumption in the snippet above is that ``CACHE_POOL`` is the third component on the manifest's components list (so its component index is 2)

Reviewer: ```suggestion
      This snippet assumes that ``CACHE_POOL`` is the third component on the manifest's components list (so its component index is 2)
```

Testing the application with external flash support
===================================================

1. |open_terminal_window_with_environment|

#. Build and flash the application by completing the following commands:

   .. code-block:: console

>       west build ./ -b nrf54h20dk/nrf54h20/cpuapp -T <configuration_name>

Reviewer 2: I suggest to make this build pristine, using it will prevent build errors after code revision update, i.e. "-p"

Author: I think `-p` rarely appears in the documentation, I'd rather not add it as it is not strictly necessary

Reviewer 2: There are other places with use of pristine, it helps avoid build errors and is a better example, I encountered problems on my desk when performing those steps without pristine.
